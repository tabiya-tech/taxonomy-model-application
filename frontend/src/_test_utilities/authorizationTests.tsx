import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
