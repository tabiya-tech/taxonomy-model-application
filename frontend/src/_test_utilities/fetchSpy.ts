export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
