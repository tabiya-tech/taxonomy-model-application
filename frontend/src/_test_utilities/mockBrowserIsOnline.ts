import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
