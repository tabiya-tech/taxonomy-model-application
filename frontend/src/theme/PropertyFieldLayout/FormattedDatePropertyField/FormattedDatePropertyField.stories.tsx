import { Meta, StoryObj } from "@storybook/react";
import FormattedDatePropertyField from "./FormattedDatePropertyField";

const meta: Meta = {
  title: "components/FormattedDatePropertyField",
  component: FormattedDatePropertyField,
  tags: ["autodocs"],
  args: {
    label: "Date",
    date: new Date(),
  },
};

type Story = StoryObj<typeof FormattedDatePropertyField>;

export const Shown: Story = {};

export default meta;
