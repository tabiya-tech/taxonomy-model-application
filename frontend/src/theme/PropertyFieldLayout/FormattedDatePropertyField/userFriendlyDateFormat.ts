const options: Intl.DateTimeFormatOptions = {
  weekday: "short", // "Mon" through "Sun"
  year: "numeric",
  month: "short", // "Jan" through "Dec"
  day: "numeric",
  hour: "numeric",
  minute: "numeric",
  hour12: true, // Use 12-hour time format, use `false` for 24-hour format
};

export const formatDate = (date: Date) => new Intl.DateTimeFormat("en-us", options).format(date);
