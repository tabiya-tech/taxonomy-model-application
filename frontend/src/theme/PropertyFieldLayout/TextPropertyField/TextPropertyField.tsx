import { Typography } from "@mui/material";
import React from "react";
import PropertyFieldLayout from "src/theme/PropertyFieldLayout/PropertyFieldLayout";

export interface TextPropertyFieldProps {
  label: string;
  text: string;
  "data-testid"?: string;
  fieldId: string;
}

const uniqueId = "5abb0d55-4073-4f5b-8432-ee8c96aa4df2";
export const DATA_TEST_ID = {
  TEXT_PROPERTY_FIELD_TEXT: `text-property-field-text-${uniqueId}`,
};

const TextPropertyField = (props: Readonly<TextPropertyFieldProps>) => {
  return (
    <PropertyFieldLayout title={props.label} data-testid={props["data-testid"]} fieldId={props.fieldId}>
      <Typography
        variant="body1"
        display="flex"
        justifyContent="space-between"
        alignItems="center"
        data-testid={DATA_TEST_ID.TEXT_PROPERTY_FIELD_TEXT}
        role="text"
      >
        {props.text}
      </Typography>
    </PropertyFieldLayout>
  );
};

export default TextPropertyField;
