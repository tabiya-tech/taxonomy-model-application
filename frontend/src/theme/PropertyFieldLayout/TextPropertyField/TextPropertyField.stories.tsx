import { Meta, StoryObj } from "@storybook/react";
import { getRandomLorem } from "src/_test_utilities/specialCharacters";
import TextPropertyField from "./TextPropertyField";

const meta: Meta = {
  title: "components/TextPropertyField",
  component: TextPropertyField,
  tags: ["autodocs"],
  args: {
    label: "foo",
    text: getRandomLorem(100),
  },
};

type Story = StoryObj<typeof TextPropertyField>;

export const Shown: Story = {};

export const ShownWithLongText: Story = {
  args: {
    text: getRandomLorem(1000),
  },
};

export default meta;
