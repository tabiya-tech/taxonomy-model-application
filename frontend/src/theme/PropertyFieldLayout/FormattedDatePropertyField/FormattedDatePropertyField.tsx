import React from "react";
import TextPropertyField from "src/theme/PropertyFieldLayout/TextPropertyField/TextPropertyField";
import { formatDate } from "src/theme/PropertyFieldLayout/FormattedDatePropertyField/userFriendlyDateFormat";

interface FormattedDatePropertyFieldProps {
  label: string;
  date: Date;
  "data-testid"?: string;
  fieldId: string;
}

const FormattedDatePropertyField = (props: Readonly<FormattedDatePropertyFieldProps>) => {
  return (
    <TextPropertyField
      label={props.label}
      text={formatDate(props.date)}
      fieldId={props.fieldId}
      data-testid={props["data-testid"]}
    />
  );
};

export default FormattedDatePropertyField;
