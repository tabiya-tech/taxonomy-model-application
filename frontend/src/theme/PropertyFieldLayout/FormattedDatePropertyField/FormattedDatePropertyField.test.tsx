// mute the console
import "src/_test_utilities/consoleMock";

import FormattedDatePropertyField from "./FormattedDatePropertyField";
import { render, screen } from "@testing-library/react";
import TextPropertyField from "src/theme/PropertyFieldLayout/TextPropertyField/TextPropertyField";

// Mock the TextPropertyField component
jest.mock("src/theme/PropertyFieldLayout/TextPropertyField/TextPropertyField", () => {
  return {
    __esModule: true,
    default: jest.fn().mockImplementation((props) => {
      return <div data-testid={props["data-testid"]}> </div>;
    }),
  };
});

describe("FormattedDatePropertyField", () => {
  test("renders correctly", () => {
    // GIVEN a Date object
    const givenDate = new Date(2000, 0, 1, 6, 0, 0);
    // AND a label
    const givenLabel = "foo-label";
    // AND a data-testid
    const givenDataTestId = "foo-data-test-id";
    // AND a fieldId
    const givenFieldId = "foo-field-id";
    // AND the userFriendlyDateFormat function will return a given string formatted date
    // to avoid breaking the snapshot test as the date will change depending on the local timezone
    const givenUserFriendlyFormattedDate = "foo-date-user-friendly-formatted";
    const formatDateSpy = jest
      .spyOn(require("src/theme/PropertyFieldLayout/FormattedDatePropertyField/userFriendlyDateFormat"), "formatDate")
      .mockReturnValue(givenUserFriendlyFormattedDate);

    // WHEN the component is rendered
    render(
      <FormattedDatePropertyField
        date={givenDate}
        label={givenLabel}
        data-testid={givenDataTestId}
        fieldId={givenFieldId}
      />
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the component to be displayed
    const formattedDatePropertyComponent = screen.getByTestId(givenDataTestId);
    expect(formattedDatePropertyComponent).toBeInTheDocument();

    // AND the formatDateSpy function to have been called with the given date
    expect(formatDateSpy).toHaveBeenCalledWith(givenDate);
    // AND the TextPropertyField to have been called with the
    // givenLabel, givenUserFriendlyFormattedDate, givenFieldId, and givenDataTestId
    const expectedTextPropertyFieldProps = {
      label: givenLabel,
      text: givenUserFriendlyFormattedDate,
      fieldId: givenFieldId,
      "data-testid": givenDataTestId,
    };
    expect(TextPropertyField).toHaveBeenCalledWith(expectedTextPropertyFieldProps, {});
    // AND to match the snapshot
    expect(formattedDatePropertyComponent).toMatchSnapshot();
  });
});
