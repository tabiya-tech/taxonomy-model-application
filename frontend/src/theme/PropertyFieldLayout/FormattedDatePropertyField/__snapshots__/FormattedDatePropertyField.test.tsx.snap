// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedDatePropertyField renders correctly 1`] = `
<div
  data-testid="foo-data-test-id"
>
   
</div>
`;
